Libraries/LibWeb/TrustedTypes/TrustedTypePolicyFactory.cpp

```diff
@@ -0,0 +1,122 @@
+
+GC::Ref<TrustedTypePolicyFactory> TrustedTypePolicyFactory::create(JS::Realm& realm)
+{
+    return realm.create<TrustedTypePolicyFactory>(realm);
+}
+
+// https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-getattributetype
+Optional<String> TrustedTypePolicyFactory::get_attribute_type(String const& tag_name, String& attribute, Optional<String> element_ns, Optional<String> attr_ns)
+{
+    // 1. Set localName to tagName in ASCII lowercase.
+    auto const local_name = tag_name.to_ascii_lowercase();
+
+    // 2. Set attribute to attribute in ASCII lowercase.
+    attribute = attribute.to_ascii_lowercase();
+
+    // 3. If elementNs is null or an empty string, set elementNs to HTML namespace.
+    if (!element_ns.has_value() || element_ns.value().is_empty())
+        element_ns = String { Namespace::HTML };
+
+    // 4. If attrNs is an empty string, set attrNs to null.
+    if (attr_ns.has_value() && attr_ns.value().is_empty())
+        attr_ns.clear();
+
+    // FIXME: We don't have a method in ElementFactory that can give us the interface name but these are all the cases
+    // we care about in the table in get_trusted_type_data_for_attribute function
+    // 5. Let interface be the element interface for localName and elementNs.
+    String interface;
+    if (local_name == HTML::TagNames::iframe && element_ns == Namespace::HTML) {
+        interface = "HTMLIFrameElement"_string;
+    } else if (local_name == HTML::TagNames::script && element_ns == Namespace::HTML) {
+        interface = "HTMLScriptElement"_string;
+    } else if (local_name == SVG::TagNames::script && element_ns == Namespace::SVG) {
+        interface = "SVGScriptElement"_string;
+    } else {
+        interface = "Element"_string;
+    }
+
+    // 6. Let expectedType be null.
+    Optional<String> expected_type {};
+
+    // 7. Set attributeData to the result of Get Trusted Type data for attribute algorithm,
+    // with the following arguments, interface as element, attribute, attrNs
+    auto const attribute_data = get_trusted_type_data_for_attribute(interface, attribute, attr_ns);
+
+    // 8. If attributeData is not null, then set expectedType to the interface’s name of the value of the fourth member of attributeData.
+    if (attribute_data.has_value()) {
+        expected_type = attribute_data.value().trusted_type;
+    }
+
+    // 9. Return expectedType.
+    return expected_type;
+}
+
+TrustedTypePolicyFactory::TrustedTypePolicyFactory(JS::Realm& realm)
+    : PlatformObject(realm)
+{
+}
+
+void TrustedTypePolicyFactory::initialize(JS::Realm& realm)
+{
+    WEB_SET_PROTOTYPE_FOR_INTERFACE(TrustedTypePolicyFactory);
+    Base::initialize(realm);
+}
+
+// https://w3c.github.io/trusted-types/dist/spec/#abstract-opdef-get-trusted-type-data-for-attribute
+Optional<TrustedTypeData const&> get_trusted_type_data_for_attribute(String const& element, String const& attribute, Optional<String> const& attribute_ns)
```

Comment:
It looks like the sanitizer is tripping up because you're returning a reference to a stack allocated variable here.

Reply:
Fixed it by making a copy at the last moment.

Reply:
I think you're right. This just ends up doing a copy with extra steps. My apologies.

Reply:
Running it with GDB and breaking in TrustedTypeData::TrustedTypeData show 2 copies in your scenario.

One in `data = matching_row.release_value();` & another in the `return` statement

I am going to be honest not sure exactlly why this one makes 2 copies you probally need a phd in c++ to be sure but my understanding is we need to make a copy of the data otherwise we will be handing over a reference to static variable of a function